clang/docs/BoundsSafety.rst

```diff
@@ -0,0 +1,362 @@
+
+* ``__bidi_indexable`` casts to ``__indexable`` if the pointer does not have an underflow. The compiler may insert run-time checks to ensure the pointer is not below the lower bound.
+
+* ``__indexable`` casts to ``__bidi_indexable``. The resulting ``__bidi_indexable`` gets the lower bound same as the pointer value.
+
+* A type conversion may involve both a bitcast and a bounds annotation cast. For example, casting from ``int *__bidi_indexable`` to ``char *__single`` involve a bitcast (``int *`` to ``char *``) and a bounds annotation cast (``__bidi_indexable`` to ``__single``). In this case, the compiler performs the bitcast and then converts the bounds annotation. This means, ``int *__bidi_indexable`` will be converted to ``char *__bidi_indexable`` and then to ``char *__single``.
+
+* ``__terminated_by(T)`` cannot cast to any safe pointer type without the same ``__terminated_by(T)`` attribute. To perform the cast, programmers can use an intrinsic function such as ``__unsafe_terminated_by_to_indexable(P)`` to force the conversion.
+
+* ``__terminated_by(T)`` can cast to ``__unsafe_indexable``.
+
+* Any type without ``__terminated_by(T)`` cannot cast to ``__terminated_by(T)`` without explicitly using an intrinsic function to allow it.
+
+  + ``__unsafe_terminated_by_from_indexable(T, PTR [, PTR_TO_TERM])`` casts any safe pointer PTR to a ``__terminated_by(T)`` pointer. ``PTR_TO_TERM`` is an optional argument where the programmer can provide the exact location of the terminator. With this argument, the function can skip reading the entire array in order to locate the end of the pointer (or the upper bound). Providing an incorrect ``PTR_TO_TERM`` causes a run-time trap.
+
+  + ``__unsafe_forge_terminated_by(T, P, E)`` creates ``T __terminated_by(E)`` pointer given any pointer ``P``. Tmust be a pointer type.
+
+Portability with toolchains that do not support the extension
+-------------------------------------------------------------
+
+The language model is designed so that it doesn’t alter the semantics of the original C program, other than introducing deterministic traps where otherwise the behavior is undefined and/or unsafe. The model has this property that when the extension is disabled, annotations compile to empty macros, thus the same source code compiles as a normal C program without any bounds annotations. The annotations used in this document are macro-defined as type attributes. This simplifies adoption both in Clang and other toolchains by not introducing any new keywords or altering the grammar. Toolchains not supporting this extension can simply macro-define the annotations to empty. For example, the toolchain not supporting this extension may not have a header defining ``__counted_by``, so the code using ``__counted_by`` must define it as nothing or include a header that has the define.
```

Comment:
This is the first mention in the document that these are macros and not keywords; if these are macros that are predefined by the compiler when using `-fbounds-safety`, we should make that more explicit in the front matter.

`Toolchains not supporting this extension can simply macro-define the annotations to empty.` -> this means the code is inherently not portable the way the document claims. For example, an older Clang will never "simply macro-define the annotations" to anything -- that's the user's responsibility. The same will be true of other non-Clang compilers.

In some ways, it would be nice if we could produce a *redistributable* header file that does this for users. e.g., it's a header-only "library" that defines macros and users have the rights to use and redistribute the header for their own purposes. This means the user would be required to include the header before using the annotations, but C programmers are very used to this model already. What's more, I think we have this same need for another, totally unrelated feature: thread safety annotations. We currently expose an interface via some documentation, so users can copy and paste that into their own projects. It would be better if we just shipped it as a header. However, this is also a bigger ask because we have a lot of deployment questions around the library (does it live in the `clang` directory or is this a different top-level project like libc++? what license is used? etc).

Reply:
The header with micro-defines lives in `llvm-project/clang/lib/Headers/ptrcheck.h`. We could also add link to this header (of llvm-project trunk) in the document once we add the header. WDYT?

Reply:
I tweaked the text to clarify that we provide the toolchain header that defines the macros.